### PR TITLE
Add Android "R/W External Storage" runtime permission request

### PR DESCRIPTION
**Pull request by cagdasc**
Permission request came with Android 6. Because if you need to make sensitive operation like write or read external storage, you need to ask to user for permission an make this operation. In Godot has only include camera, record audio and vibration permission. We need to write and read some data in game for save and load states. So I add write and read external storage permission requests and all permission request result callback but I am not sure results were emiting on Main Loop.

Edit:
3.1 branch PR: #31583

### CODE REVIEW

**Comment by Sslaxx** [comment]
Why did you close the previous request?

**Comment by cagdasc** [comment]
It gives clang format error and I revert the commit and fix-push again.

**Comment by cagdasc** [comment]
@akien-mga why PR is waiting for merge?

**Comment by akien-mga** [comment]
Because someone has to review it before merging :)

**Comment by cagdasc** [comment]
Okay, I suppose pr was revived 😂

**Comment by volzhs** [comment]
I'm not against about adding signal but now we seems to have to deal with all kinds of permission.
there should be a way to simply handle this.

**Comment by cagdasc** [comment]
I think we can only add permission requests easily but only request. If I add location permission in here simply but I need to write location operation base code.In this PR we only ask users for R/W permission, Godot handle file R/W for a long time.

**Review by volzhs — APPROVED** [review]

File: platform/android/java/src/org/godotengine/godot/Godot.java
Comment (akien-mga) [inline platform/android/java/src/org/godotengine/godot/Godot.java]: Shouldn't these be 4 and 5?

File: platform/android/os_android.cpp
Comment (akien-mga) [inline platform/android/os_android.cpp]: The parameters should be `p_permission` and `p_granted`.

File: core/os/main_loop.cpp
Comment (akien-mga) [inline core/os/main_loop.cpp]: Signals shouldn't be prefixed with `_`, unless you don't want them to be visible for users. But since this signal is not used for anything in this PR, I assume you add it so that users can connect to it?

File: platform/android/java/src/org/godotengine/godot/Godot.java
Comment (cagdasc) [reply platform/android/java/src/org/godotengine/godot/Godot.java]: I fixed it while resolving conflict. I will fix it.

File: core/os/main_loop.cpp
Comment (cagdasc) [reply core/os/main_loop.cpp]: Yes, users can connect to it. I fixed it.

**Review by volzhs — APPROVED** [review]

**Comment by m4gr3d** [comment]
> I'm not against about adding signal but now we seems to have to deal with all kinds of permission.
> there should be a way to simply handle this.

The list of permissions in the manifest can be retrieved at runtime via the [PackageInfo#requestedPermissions](https://developer.android.com/reference/android/content/pm/PackageInfo.html#requestedPermissions) field. So a better way to deal with arbitrary permissions would be to query the manifest at runtime, and request to the user the permissions declared there.

In addition, we should include an api to allow GDScript to request arbitrary runtime permissions, and be notified of the result via notifications. I'll create an issue to track it. It may be relatively short to implement (~1 week), so we may want to hold off on this PR if time's not an issue.

**Comment by volzhs** [comment]
I think we can wait. 😄

**Comment by m4gr3d** [comment]
@cagdasc Should this PR be closed since it'll be replaced by PR #32246?

**Pull request closed without merging**
